test/e2e/versionskew/operator.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2020 Intel Corporation.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package versionskew
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	api "github.com/intel/pmem-csi/pkg/apis/pmemcsi/v1beta1"
+	"github.com/intel/pmem-csi/test/e2e/deploy"
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = deploy.DescribeForSome("versionskew", func(d *deploy.Deployment) bool {
+	// Run these tests only for bare operator deployments, not for the operator
+	// deployed by OLM. Because there is no way how to ask olm to stop the operator
+	// but keep the CRD.
+	return d.HasOperator && !d.HasDriver && !d.HasOLM
+}, func(d *deploy.Deployment) {
+	var (
+		evWatcher  watch.Interface
+		evWait     sync.WaitGroup
+		evCaptured map[types.UID]bool
+		evMutex    sync.Mutex
+	)
+
+	f := framework.NewDefaultFramework("skew")
+	f.SkipNamespaceCreation = true
+
+	BeforeEach(func() {
+		var err error
+
+		Expect(f).ShouldNot(BeNil(), "framework init")
+
+		evWatcher, err = f.ClientSet.CoreV1().Events("").Watch(context.TODO(), metav1.ListOptions{})
+		Expect(err).ShouldNot(HaveOccurred(), "get events watcher")
+
+		evMutex = sync.Mutex{}
+	})
+
+	deployDriver := func(name, image string) api.PmemCSIDeployment {
+		cr := api.PmemCSIDeployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: api.DeploymentSpec{
+				DeviceMode: api.DeviceModeDirect,
+			},
+		}
+		if image != "" {
+			cr.Spec.Image = image
+		}
+		driverDep := deploy.CreateDeploymentCR(f, cr)
+		return driverDep
+	}
+
+	changeVersion := func(version string) {
+		newDeploymentName := d.Name()
+		if version != "" {
+			// deploy.Parse() expects the deployment is seperated with '-'
+			// and the version should be the 4th field.
+			newDeploymentName = d.Name() + "---" + version
+		}
+
+		// Change operator release.
+		By(fmt.Sprintf("Switching the operator to %s", newDeploymentName))
+		deployment, err := deploy.Parse(newDeploymentName)
+		framework.ExpectNoError(err, "internal error while parsing %s: %v", newDeploymentName, err)
+		deploy.EnsureDeploymentNow(f, deployment)
+	}
+
+	captureEvents := func(uuids ...types.UID) {
+		framework.Logf("Checking for events on: %s", uuids)
+
+		// Initialize events map
+		evMutex.Lock()
+		evCaptured = map[types.UID]bool{}
+		for _, uid := range uuids {
+			evCaptured[uid] = false
+		}
```

Comment:
This looks redundant. `evCaptured[uid]` is false when `uid` doesn't have an entry, so explicitly adding it with false as value can be skipped.

Reply:
Done.